bevy_lint/src/lints/borrow_of_reborrowable.rs

```diff
@@ -0,0 +1,224 @@
+//! Checks for function parameters that take a mutable reference to a
+//! re-borrwable type.
+//!
+//! # Motivation
+//!
+//! Mutable references to re-borrwable types can almost always be readily
+//! converted back to an owned instance of itself, albeit with an appropriately
+//! shorter lifetime.
+//!
+//! The only time this isn't true is when the function returns referenced data
+//! that is bound to the mutable reference of the re-borrowable type.
+//!
+//! # Example
+//!
+//! ```
+//! # use bevy::prelude::*;
+//! fn system(mut commands: Commands) {
+//!   helper_function(&mut commands);
+//! }
+//!
+//! fn helper_function(commands: &mut Commands) {
+//!   // ...
+//! }
+//! ```
+//!
+//! Use instead:
+//!
+//! ```
+//! # use bevy::prelude::*;
+//! fn system(mut commands: Commands) {
+//!   helper_function(commands.reborrow());
+//! }
+//!
+//! fn helper_function(mut commands: Commands) {
+//!   // ...
+//! }
+//! ```
+
+use std::ops::ControlFlow;
+
+use crate::declare_bevy_lint;
+use clippy_utils::{diagnostics::span_lint_and_sugg, ty::match_type};
+use rustc_errors::Applicability;
+use rustc_hir::{intravisit::FnKind, Body, FnDecl, Mutability};
+use rustc_lint::{LateContext, LateLintPass, Lint};
+use rustc_middle::{
+    ty::Interner,
+    ty::{Ty, TyKind, TypeVisitable, TypeVisitor},
+};
+use rustc_session::declare_lint_pass;
+use rustc_span::{def_id::LocalDefId, symbol::Ident, Span};
+
+declare_bevy_lint! {
+    pub BORROW_OF_COMMANDS,
+    PEDANTIC,
+    "parameter takes a mutable reference to `Commands` or `EntityCommands` instead of a re-borrowed instance",
+}
+
+declare_bevy_lint! {
+    pub BORROW_OF_RESOURCE,
+    PEDANTIC,
+    "parameter takes a mutable reference to `ResMut` or `NonSendMut` instead of a re-borrowed instance",
+}
+
+declare_bevy_lint! {
+    pub BORROW_OF_QUERY,
+    PEDANTIC,
+    "parameter takes a mutable reference to `Query` instead of a re-borrowed instance",
+}
+
+declare_lint_pass! {
+    BorrowOfReborrowable => [BORROW_OF_COMMANDS.lint, BORROW_OF_RESOURCE.lint, BORROW_OF_QUERY.lint]
+}
+
+impl<'tcx> LateLintPass<'tcx> for BorrowOfReborrowable {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        _: FnKind<'tcx>,
+        decl: &'tcx FnDecl<'tcx>,
+        _: &'tcx Body<'tcx>,
+        _: Span,
+        def_id: LocalDefId,
+    ) {
+        // We are already inside of the function item,
+        // so we can use `instantiate_identity` to discharge the binder
+        let fn_sig = cx.tcx.fn_sig(def_id).instantiate_identity();
```

Comment:
I'm not entirely sure I understand your reasoning for this, but `instantiate_identity()` has always confused me a bit. Could you go into more depth your reasoning on this, and are there any drawbacks of calling `instantiate_identity()`?

Reply:
This was my initial thought process:

The [docs](https://doc.rust-lang.org/stable/nightly-rustc/rustc_infer/infer/canonical/ir/struct.EarlyBinder.html#method.instantiate_identity) state the following:
> Makes the identity replacement `T0 => T0, ..., TN => TN`. Conceptually, this converts universally bound variables into placeholders when inside of a given item.
>
> For example, consider `for<T> fn foo<T>(){ .. }`:
>
> Outside of `foo`, `T` is bound (represented by the presence of `EarlyBinder`).
Inside of the body of `foo`, we treat `T` as a placeholder by calling `instantiate_identity` to discharge the `EarlyBinder`.

So because we're in the body (I think [parameters count](https://doc.rust-lang.org/stable/nightly-rustc/rustc_hir/struct.Body.html#examples)), it should be safe to do this.

---

Upon reading this back, though, I don't think this is something that I really understood when writing that comment. My new understanding is that this doesn't actually have a chance of failing because all it's doing is filling any generics in with placeholders. And that's fine, we don't care about the types of those generics for this particular lint.

So maybe I should at the very least rephrase this comment to indicate that we're using  `instantiate_identity` since we don't care to fill them in with any real type data.

Am I completely off? 😅

Reply:
> So maybe I should at the very least rephrase this comment to indicate that we're using `instantiate_identity` since we don't care to fill them in with any real type data.

That seems correct, though interestingly enough the source of [`instantiate_identity()`](https://doc.rust-lang.org/stable/nightly-rustc/src/rustc_type_ir/binder.rs.html#589) and [`skip_binder()`](https://doc.rust-lang.org/stable/nightly-rustc/src/rustc_type_ir/binder.rs.html#408) are the same.